precompiles/erc20/integration_test.go

```diff
@@ -0,0 +1,1048 @@
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				s.expectSendAuthz(grantee.AccAddr, granter.AccAddr, bondCoins.Add(increaseCoins...))
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+			)
+
+			DescribeTable("decreasing the allowance should return an error", func(callType int) {
+				decreaseCoins := sdk.Coins{sdk.NewInt64Coin(s.tokenDenom, 100)}
+
+				txArgs, decreaseArgs := s.getTxAndCallArgs(callType, contractAddr)
+				decreaseArgs.MethodName = auth.DecreaseAllowanceMethod
+				decreaseArgs.Args = []interface{}{grantee.Addr, decreaseCoins[0].Amount.BigInt()}
+
+				notFoundCheck := execRevertedCheck
+				if callType == directCall {
+					notFoundCheck = failCheck.WithErrContains(
+						fmt.Sprintf(auth.ErrAuthzDoesNotExistOrExpired, erc20.SendMsgURL, grantee.Addr.String()),
+					)
+				}
+
+				_, _, err = s.callContractAndCheckLogs(granter.Priv, txArgs, decreaseArgs, notFoundCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+			)
+		})
+
+		When("an allowance exists for the same token", func() {
+			var authzCoins sdk.Coins
+
+			BeforeEach(func() {
+				authzCoins = sdk.NewCoins(
+					sdk.NewInt64Coin(s.network.GetDenom(), 100),
+					sdk.NewInt64Coin(s.tokenDenom, 200),
+				)
+
+				s.setupSendAuthz(grantee.AccAddr, granter.Priv, authzCoins)
+			})
+
+			DescribeTable("increasing the allowance should increase the spend limit", func(callType int) {
+				increaseCoins := sdk.Coins{sdk.NewInt64Coin(s.tokenDenom, 100)}
+
+				txArgs, increaseArgs := s.getTxAndCallArgs(callType, contractAddr)
+				increaseArgs.MethodName = auth.IncreaseAllowanceMethod
+				increaseArgs.Args = []interface{}{grantee.Addr, increaseCoins[0].Amount.BigInt()}
+
+				approveCheck := passCheck.WithExpEvents(auth.EventTypeApproval)
+
+				_, _, err = s.callContractAndCheckLogs(granter.Priv, txArgs, increaseArgs, approveCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				s.expectSendAuthz(grantee.AccAddr, granter.AccAddr, authzCoins.Add(increaseCoins...))
+			},
+				Entry(" - direct call", directCall),
+				// FIXME: this also shows interesting behavior because when calling this there is an authorization from the contract to the grantee
+				// instead of the increasing the one from granter to the grantee, because the granter is always taken as the contract caller (in this case the smart contract),
+				// even though we sign with the granter key. I think this is different to how we have it implemented for other precompiles, e.g. staking?
+				//
+				// See IncreaseAllowance method in the approve.go file:
+				//
+				// ```
+				// granter := contract.CallerAddress
+				// ```
+				Entry(" - through contract", contractCall),
```

Comment:
The flow here is a little counterintuitive / different from how we have it implemented in the other precompiles.  Here, when approving a spend limit from granter to grantee through a smart contract, what's actually created is an authorization from the smart contract to the grantee. 🤔

Reply:
ok this needs to be fixed then

Reply:
This is the same flow as it is present on the original ERC20 implementation - even though it's different from our other precompiles, we will stick with it to have full ERC20 compatibility/alignment (for now).